utils/get_prometheus_data.py

```diff
@@ -0,0 +1,145 @@
+            self.headers = {'Authorization': bearer}
+        else:
+            logger.warn("""snafu service account token and prometheus url not set \n
+                        No Prometheus data will be indexed""")
+
+    def api_call(self, query):
+
+        api_call = self.url + query
+
+        response = requests.get(api_call,
+                                headers=self.headers,
+                                verify=False)
+
+        if "success" in response.json()["status"]:
+            return response.json()["data"]
+        else:
+            raise Exception(response["status"])
+
+    def get_label_list(self):
+
+        query = "/api/v1/label/__name__/values"
+
+        try:
+            results = self.api_call(query)
+            return results
+        except Exception as e:
+            logger.warn("failure to get labels %s" % e)
+
+    def get_all_metrics(self):
+
+        # check get_data bool, if false by-pass all processing
+        if self.get_data:
+            start_time = time.time()
+
+            filename = os.path.join(sys.path[0], 'utils/prometheus_labels/included_labels.json')
+            with open(filename, 'r') as f:
+                datastore = json.load(f)
+
+            for label in self.get_label_list():
+
+                if label in datastore["data"]:
+
+                    query = "/api/v1/query_range?query=" + label
+                    query = query + "&start=" + self.start
+                    query = query + "&end=" + self.end
+                    query = query + "&step=" + str(self.T_Delta) + "s"
+                    # query_start_time = time.time()
+                    try:
+                        response = self.api_call(query)
+                    except Exception as e:
+                        logger.warn("failure to get metric results %s" % e)
+
+                    results = response['result']
+
+                    # results is a list of all hits
+
+                    for result in results:
+                        # clean up name key from __name__ to name
+                        result["metric"]["name"] = ""
+                        result["metric"]["name"] = result["metric"]["__name__"]
+                        del result["metric"]["__name__"]
+                        ROC_V1 = 0
+                        # each result has a list, we must flatten it out in order to send to ES
+                        for value in result["values"]:
+                            # fist index is time stamp
+                            timestamp = datetime.utcfromtimestamp(value[0]).strftime('%Y-%m-%dT%H:%M:%S.%fZ')
+                            # second index is value of metric
+                            if "NaN" in value[1]:  # need to handle values that are NaN, Inf, or -Inf
+                                metric_value = 0
+                            else:
+                                """
+                                TODO add logic for calculating the rate of increase per second
```

Comment:
don't understand comments.  So if prometheus restarts, counters all go to zero after some delay and we have to compute results differently, how does that work?   Can prometheus counters ever wrap (i.e. 32-bit counters often do)?   And why a class variable?    Why is T-delta always 30 sec?   Can't we get that from just looking at current timestamp - previous timestamp?

Reply:
Prometheus will reset counts on its own, the formula prometheus provided was to detect a value smaller than the current and reset it accordingly.